includes/validation/class-amp-site-validation.php

```diff
@@ -0,0 +1,279 @@
+<?php
+/**
+ * Class AMP_Site_Validation
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Site_Validation
+ *
+ * @since 1.0
+ */
+class AMP_Site_Validation {
+
+	/**
+	 * The size of the batch of URLs to validate.
+	 * Grouped in a batch to avoid making loopback requests for all of a site's URLs at the same time.
+	 *
+	 * @var int
+	 */
+	const BATCH_SIZE = 100;
+
+	/**
+	 * The WP-CLI argument to validate the site.
+	 * The full command is: wp amp validate-site.
+	 *
+	 * @var int
+	 */
+	const WP_CLI_ARGUMENT = 'validate-site';
+
+	/**
+	 * The WP CLI progress bar.
+	 *
+	 * @see https://make.wordpress.org/cli/handbook/internal-api/wp-cli-utils-make-progress-bar/
+	 * @var cli\progress\Bar|WP_CLI\NoOp
+	 */
+	public static $wp_cli_progress;
+
+	/**
+	 * The total number of validation errors, regardless of whether they were accepted.
+	 *
+	 * @var int
+	 */
+	public static $total_errors = 0;
+
+	/**
+	 * The total number of unaccepted validation errors.
+	 *
+	 * If an error has been accepted in the /wp-admin validation UI,
+	 * it won't count toward this.
+	 *
+	 * @var int
+	 */
+	public static $unaccepted_errors = 0;
+
+	/**
+	 * The number of URLs crawled, regardless of whether they have validation errors.
+	 *
+	 * @var int
+	 */
+	public static $number_crawled = 0;
+
+	/**
+	 * Inits the class.
+	 */
+	public static function init() {
+		if ( defined( 'WP_CLI' ) && WP_CLI ) {
+			WP_CLI::add_command( 'amp', array( __CLASS__, 'crawl_site' ) );
+		}
+	}
+
+	/**
+	 * Crawls the entire site to validate it, and gets the results.
+	 *
+	 * @param array $args The arguments for the command.
+	 */
+	public static function crawl_site( $args ) {
```

Comment:
This and `init()` are the only methods in the class that aren't tested, as they depend on the class `WP_CLI`.